tests/Http/Plugin/ExceptionPluginTest.php

```diff
@@ -0,0 +1,77 @@
+<?php
+declare(strict_types=1);
+
+namespace Lmc\Matej\Http\Plugin;
+
+use Fig\Http\Message\StatusCodeInterface;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Psr7\Response;
+use Http\Client\Promise\HttpFulfilledPromise;
+use Lmc\Matej\Exception\AuthorizationException;
+use Lmc\Matej\Exception\RequestException;
+use PHPUnit\Framework\TestCase;
+use Psr\Http\Message\RequestInterface;
+
+class ExceptionPluginTest extends TestCase
+{
+    /** @var callable */
+    protected $emptyFunction;
+
+    /** @before */
+    public function init(): void
+    {
+        $this->emptyFunction = function (): void {
+        };
+    }
+
+    /** @test */
+    public function shouldReturnResponseWhenNoError(): void
+    {
+        $request = new Request('GET', 'http://foo.com/endpoint');
+        $response = new Response(StatusCodeInterface::STATUS_OK);
+
+        $next = function (RequestInterface $receivedRequest) use ($request, $response) {
+            $this->assertSame($request, $receivedRequest);
+
+            return new HttpFulfilledPromise($response);
+        };
+
+        $plugin = new ExceptionPlugin();
+        $promise = $plugin->handleRequest($request, $next, $this->emptyFunction);
+        $this->assertInstanceOf(HttpFulfilledPromise::class, $promise);
+    }
+
+    /**
+     * @test
+     * @dataProvider provideErrorStatusCodes
+     */
+    public function shouldThrowExceptionBasedOnStatusCode(int $statusCode, string $expectedExceptionClass): void
+    {
+        $request = new Request('GET', 'http://foo.com/endpoint');
+        $response = new Response($statusCode);
+
+        $next = function (RequestInterface $receivedRequest) use ($request, $response) {
+            $this->assertSame($request, $receivedRequest);
+
+            return new HttpFulfilledPromise($response);
+        };
+
+        $plugin = new ExceptionPlugin();
+
+        $this->expectException($expectedExceptionClass);
+        $plugin->handleRequest($request, $next, $this->emptyFunction);
+    }
+
+    /**
+     * @return array[]
+     */
+    public function provideErrorStatusCodes(): array
+    {
+        return [
+            'HTTP 400' => [StatusCodeInterface::STATUS_BAD_REQUEST, RequestException::class],
+            'HTTP 401' => [StatusCodeInterface::STATUS_UNAUTHORIZED, AuthorizationException::class],
+            'HTTP 404' => [StatusCodeInterface::STATUS_NOT_FOUND, RequestException::class],
+            'HTTP 500' => [StatusCodeInterface::STATUS_INTERNAL_SERVER_ERROR, RequestException::class],
```

Comment:
What about testing `< 600` boundary.

Reply:
There is no HTTP 600 :). Byt I added the check for (imaginary) HTTP 599 code as well.